File: kinv/core/inventory/fields.py
"""The schematic fields a resolved part is worth writing back."""

from __future__ import annotations

from typing import Iterable


def fields_for_part(resolution: dict) -> dict[str, str]:
    """Only what the tool actually decided: the part number, who makes it, where to buy it.

    ``Value`` and ``Footprint`` are the designer's; ``Description`` and
    ``Datasheet`` come from the symbol library and overwriting them would
    replace something a person chose with something a distributor said.
    """
    part = resolution.get("part")
    if part is None:
        return {}
    fields = {"MPN": part["mpn"]}
    if part.get("manufacturer", "") != "":
        fields["Manufacturer"] = part["manufacturer"]
    # Supplier-neutral names: a field called `DigiKey#` would say the board is
    # bought from one distributor forever.
    if part.get("supplier") is not None:
        fields["Supplier"] = part["supplier"]
    if part.get("orderNumber") is not None:
        fields["Supplier#"] = part["orderNumber"]
    return fields


def fields_to_write(parts: Iterable[dict], resolutions: Iterable[dict]) -> list[dict]:
    """Every assigned part's fields, against the placements that carry them."""
    by_key = {r["key"]: r for r in resolutions}
    entries = []
    for part in parts:
        resolution = by_key.get(part["key"])
        if resolution is None:
            continue
        fields = fields_for_part(resolution)
        if not fields:
            continue
        # Every spelling that fed this part: 100n and 100nF are one purchase.
        values = list(dict.fromkeys(source["value"] for source in part["sources"]))
        entries.append({"key": part["key"], "refs": part["refs"], "values": values, "fields": fields})
    return entries

File: src/core/inventory/fields.ts
import type { PartRollup } from "../consolidate/findings.js";
import type { Resolution } from "./types.js";

/**
 * The schematic fields a resolved part is worth writing back.
 *
 * Only what the tool actually decided. `Value` and `Footprint` are the
 * designer's; `Description` and `Datasheet` come from the symbol library and
 * overwriting them would replace something a person chose with something a
 * distributor said. What is left is the part number, who makes it, and where
 * to buy it — the three facts a BOM export needs and a schematic cannot
 * currently tell anyone.
 */
export function fieldsForPart(resolution: Resolution): Record<string, string> {
  const part = resolution.part;
  if (part === undefined) return {};

  const fields: Record<string, string> = { MPN: part.mpn };
  if (part.manufacturer !== "") fields["Manufacturer"] = part.manufacturer;
  // Supplier-neutral names: the board is not bought from one distributor
  // forever, and a field called `DigiKey#` would say otherwise.
  if (part.supplier !== undefined) fields["Supplier"] = part.supplier;
  if (part.orderNumber !== undefined) fields["Supplier#"] = part.orderNumber;
  return fields;
}

export type FieldPlanEntry = {
  key: string;
  refs: string[];
  /** The spellings this part's BOM lines used, so the writer can check it found the right symbol. */
  values: string[];
  fields: Record<string, string>;
};

/** Every assigned part's fields, against the placements that carry them. */
export function fieldsToWrite(
  parts: readonly PartRollup[],
  resolutions: readonly Resolution[],
): FieldPlanEntry[] {
  const byKey = new Map(resolutions.map((r) => [r.key, r]));
  return parts.flatMap((part) => {
    const resolution = byKey.get(part.key);
    if (resolution === undefined) return [];
    const fields = fieldsForPart(resolution);
    if (Object.keys(fields).length === 0) return [];
    // Every spelling that fed this part: 100n and 100nF are one purchase, and
    // either is the right symbol to write the part number onto.
    const values = [...new Set(part.sources.map((s) => s.value))];
    return [{ key: part.key, refs: part.refs, values, fields }];
  });
}
